### PR TITLE
Simplify stats calculation code

### PR DESCRIPTION
**Pull request by findepi**

### CODE REVIEW

File: presto-main/src/main/java/com/facebook/presto/cost/CoefficientBasedStatsCalculator.java [hunk 1]
```diff
@@ -195,12 +180,12 @@ private PlanNodeStatsEstimate visitTableScanWithPredicate(TableScanNode node, Ex
             return tableScanStats;
         }
 
-        private Constraint<ColumnHandle> getConstraint(TableScanNode node, Expression predicate)
+        private Constraint<ColumnHandle> getConstraint(TableScanNode node)
         {
             DomainTranslator.ExtractionResult decomposedPredicate = DomainTranslator.fromPredicate(
```
Comment (rschlussel2) [inline presto-main/src/main/java/com/facebook/presto/cost/CoefficientBasedStatsCalculator.java]: Is the decomposedPredicate still needed if the predicate is always TRUE_LITERAL?  Seems like this whole function can be replaced with `node.getCurrentConstraint`.

File: presto-main/src/main/java/com/facebook/presto/cost/CoefficientBasedStatsCalculator.java [hunk 2]
```diff
@@ -117,17 +116,8 @@ public PlanNodeStatsEstimate visitOutput(OutputNode node, Void context)
...
-                sourceStats = visitTableScanWithPredicate((TableScanNode) node.getSource(), node.getPredicate());
-            }
-            else {
-                sourceStats = visitSource(node);
-            }
-
-            final double filterCoefficient = FILTER_COEFFICIENT;
-            PlanNodeStatsEstimate filterStats = sourceStats
-                    .mapOutputRowCount(value -> value * filterCoefficient);
+            PlanNodeStatsEstimate sourceStats = visitSource(node);
```
Comment (rschlussel2) [inline presto-main/src/main/java/com/facebook/presto/cost/CoefficientBasedStatsCalculator.java]: To make sure I understand.  We added complicated logic here because we didn't used to have the table layouts yet when the stats calculator was running.  But now that we run PickTableLayout earlier, we don't need this.

File: presto-main/src/main/java/com/facebook/presto/cost/CoefficientBasedStatsCalculator.java [hunk 3]
```diff
@@ -136,11 +135,8 @@ public PlanNodeStatsEstimate visitJoin(JoinNode node, Void context)
             PlanNodeStatsEstimate rightStats = sourceStats.get(1);
 
             PlanNodeStatsEstimate.Builder joinStats = PlanNodeStatsEstimate.builder();
-            if (!leftStats.getOutputRowCount().isValueUnknown() && !rightStats.getOutputRowCount().isValueUnknown()) {
-                double rowCount = Math.max(leftStats.getOutputRowCount().getValue(), rightStats.getOutputRowCount().getValue()) * JOIN_MATCHING_COEFFICIENT;
-                joinStats.setOutputRowCount(new Estimate(rowCount));
-            }
-
+            double rowCount = Math.max(leftStats.getOutputRowCount(), rightStats.getOutputRowCount()) * JOIN_MATCHING_COEFFICIENT;
```
Comment (rschlussel2) [inline presto-main/src/main/java/com/facebook/presto/cost/CoefficientBasedStatsCalculator.java]: In the commit message title:
Use doubles instead *of* Estimates in PlanNodeStatsEstimate

And at the end of the commit message "connectors" is spelled incorrectly (there's an extra 'e')

File: presto-main/src/main/java/com/facebook/presto/cost/CoefficientBasedStatsCalculator.java (hunk 2 above)
Comment (findepi) [reply presto-main/src/main/java/com/facebook/presto/cost/CoefficientBasedStatsCalculator.java]: This is correct. `TestCoefficientBasedStatsCalculator.testStatsCalculatorUsesLayout` still passes, showing the logic is no longer necessary.

File: presto-main/src/main/java/com/facebook/presto/cost/CoefficientBasedStatsCalculator.java (hunk 1 above)
Comment (findepi) [reply presto-main/src/main/java/com/facebook/presto/cost/CoefficientBasedStatsCalculator.java]: Theoretically there might be the case that not the whole `node.getCurrentConstraint()` is used (see `decomposedPredicate.getRemainingExpression()`). I can't imagine this ever being the case (normally, we wouldn't store such a constraint in the `TableScanNode` in the first place), but I wouldn't hard-code this assumption here.

Comment (rschlussel2) [reply presto-main/src/main/java/com/facebook/presto/cost/CoefficientBasedStatsCalculator.java]: I don't see how this helps:
`decomposedPredicate` will be an `ExtractionResult` of `TupleDomain.all()`.
`decomposedPredicate.getTupleDomain()` will give you `TupleDomain.all()`
`transform` does nothing because the domains was already an empty map
and then `intersect(node.getCurrentConstraint()` will just give you the `currentConstraint`.  

Am I missing something?

Comment (findepi) [reply presto-main/src/main/java/com/facebook/presto/cost/CoefficientBasedStatsCalculator.java]: I finally got that. You're right.

File: presto-tpch/src/main/java/com/facebook/presto/tpch/TpchMetadata.java [hunk 4]
```diff
@@ -408,7 +408,9 @@ public ColumnMetadata getColumnMetadata(ConnectorSession session, ConnectorTable
 
     private Predicate<NullableValue> convertToPredicate(TupleDomain<ColumnHandle> predicate, TpchColumn column)
     {
-        return nullableValue -> predicate.contains(TupleDomain.fromFixedValues(ImmutableMap.of(toColumnHandle(column), nullableValue)));
+        TpchColumnHandle columnHandle = toColumnHandle(column);
+        TupleDomain<ColumnHandle> columnPredicate = filterColumns(predicate, columnHandle::equals);
+        return nullableValue -> columnPredicate.contains(TupleDomain.fromFixedValues(ImmutableMap.of(columnHandle, nullableValue)));
```
Comment (kokosing) [inline presto-tpch/src/main/java/com/facebook/presto/tpch/TpchMetadata.java]: it took me a while to understand this issue. Maybe we could provide tests for that. Or we could leave https://github.com/prestodb/presto/pull/9823 to cover this with ATQ

File: presto-main/src/test/java/com/facebook/presto/cost/TestCoefficientBasedStatsCalculator.java [hunk 5]
```diff
@@ -56,26 +55,24 @@ public void testStatsCalculatorUsesLayout()
...
+                                .withStats(PlanNodeStatsEstimate.builder()
+                                        .setOutputRowCount(new Estimate(15000.0))
+                                        .setOutputSizeInBytes(unknownValue())
+                                        .build())));
     }
 
     private void assertPlan(String sql, PlanMatchPattern pattern)
     {
-        assertPlan(sql, LogicalPlanner.Stage.CREATED, pattern);
+        assertPlan(sql, LogicalPlanner.Stage.OPTIMIZED_AND_VALIDATED, pattern);
```
Comment (kokosing) [inline presto-main/src/test/java/com/facebook/presto/cost/TestCoefficientBasedStatsCalculator.java]: it looks unrelated

File: presto-main/src/main/java/com/facebook/presto/cost/PlanNodeStatsEstimate.java [hunk 6]
```diff
@@ -63,10 +73,39 @@ public PlanNodeStatsEstimate mapOutputSizeInBytes(Function<Double, Double> mappi
         return buildFrom(this).setOutputSizeInBytes(mappingFunction.apply(outputRowCount)).build();
     }
 
+    public PlanNodeStatsEstimate mapSymbolColumnStatistics(Symbol symbol, Function<SymbolStatsEstimate, SymbolStatsEstimate> mappingFunction)
+    {
+        return buildFrom(this)
+                .setSymbolStatistics(symbolStatistics.entrySet().stream()
+                        .collect(Collectors.toMap(
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/cost/PlanNodeStatsEstimate.java]: to `toImmutableMap`?

File: presto-main/src/main/java/com/facebook/presto/cost/SymbolStatsEstimate.java [hunk 7]
```diff
@@ -0,0 +1,185 @@
...
+    {
+        checkArgument(
+                lowValue <= highValue || (isNaN(lowValue) && isNaN(highValue)),
+                "low value must be less than or equal to high value or both values have to be NaN, got %s and %s respectively",
+                lowValue,
+                highValue);
+        this.lowValue = lowValue;
+        this.highValue = highValue;
+
+        checkArgument((0 <= nullsFraction && nullsFraction <= 1.) || isNaN(nullsFraction), "Nulls fraction should be within [0, 1] or NaN, got: %s", nullsFraction);
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/cost/SymbolStatsEstimate.java]: format this in several lines as the check above

File: presto-main/src/main/java/com/facebook/presto/cost/SymbolStatsEstimate.java [hunk 8]
```diff
@@ -0,0 +1,185 @@
...
+        this.nullsFraction = nullsFraction;
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/cost/SymbolStatsEstimate.java]: maybe you should set `nullsFraction = 1.0` in case of empty range

File: presto-main/src/main/java/com/facebook/presto/cost/SymbolStatsEstimate.java [hunk 9]
```diff
@@ -0,0 +1,185 @@
...
+    {
+        if (isRangeEmpty()) {
+            return 1.0;
+        }
+        return nullsFraction;
+    }
+
+    public double getValuesFraction()
+    {
+        return 1.0 - nullsFraction;
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/cost/SymbolStatsEstimate.java]: `1.0 - getNullsFraction()`?

File: presto-main/src/main/java/com/facebook/presto/cost/SymbolStatsEstimate.java [hunk 10]
```diff
@@ -0,0 +1,185 @@
...
+        this.highValue = highValue;
+
+        checkArgument((0 <= nullsFraction && nullsFraction <= 1.) || isNaN(nullsFraction), "Nulls fraction should be within [0, 1] or NaN, got: %s", nullsFraction);
+        this.nullsFraction = nullsFraction;
+
+        checkArgument(averageRowSize >= 0 || isNaN(averageRowSize), "Average row size should be non-negative or NaN, got: %s", averageRowSize);
+        this.averageRowSize = averageRowSize;
+
+        checkArgument(distinctValuesCount >= 0 || isNaN(distinctValuesCount), "Distinct values count should be non-negative, got: %s", distinctValuesCount);
+        this.distinctValuesCount = distinctValuesCount;
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/cost/SymbolStatsEstimate.java]: maybe you should set it to 0 in case of empty range

File: presto-main/src/main/java/com/facebook/presto/cost/SymbolStatsEstimate.java [hunk 11]
```diff
@@ -0,0 +1,185 @@
...
+    }
+
+    public boolean isRangeEmpty()
+    {
+        return isNaN(lowValue) && isNaN(highValue);
+    }
+
+    public double getNullsFraction()
+    {
+        if (isRangeEmpty()) {
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/cost/SymbolStatsEstimate.java]: it is weird that this is the only case where `isRangeEmpty` is checked. What about values fraction, distinct values count, average size?

File: presto-main/src/test/java/com/facebook/presto/cost/TestCoefficientBasedStatsCalculator.java (hunk 5 above)
Comment (losipiuk) [reply presto-main/src/test/java/com/facebook/presto/cost/TestCoefficientBasedStatsCalculator.java]: I think it is. We must ensure predicate is pushed down to TS node.

File: presto-main/src/main/java/com/facebook/presto/cost/PlanNodeStatsEstimate.java (hunk 6 above)
Comment (losipiuk) [reply presto-main/src/main/java/com/facebook/presto/cost/PlanNodeStatsEstimate.java]: static import

File: presto-main/src/main/java/com/facebook/presto/cost/PlanNodeStatsEstimate.java [hunk 12]
```diff
@@ -79,14 +118,15 @@ public boolean equals(Object o)
             return false;
         }
         PlanNodeStatsEstimate that = (PlanNodeStatsEstimate) o;
-        return Objects.equals(outputRowCount, that.outputRowCount) &&
-                Objects.equals(outputSizeInBytes, that.outputSizeInBytes);
+        return Double.compare(that.outputRowCount, outputRowCount) == 0 &&
```
Comment (losipiuk) [inline presto-main/src/main/java/com/facebook/presto/cost/PlanNodeStatsEstimate.java]: Change from `Object.equals` to `Double.compare` should be part of previous commit.

File: presto-main/src/main/java/com/facebook/presto/cost/PlanNodeStatsEstimate.java [hunk 13]
```diff
@@ -31,17 +31,15 @@
 public class PlanNodeStatsEstimate
 {
     public static final PlanNodeStatsEstimate UNKNOWN_STATS = builder().build();
+    public static final double DEFAULT_DATA_SIZE_PER_COLUMN = 10;
```
Comment (losipiuk) [inline presto-main/src/main/java/com/facebook/presto/cost/PlanNodeStatsEstimate.java]: Maybe change commit message to 

```
Compute PlanNodes output size based on symbol stats
```

File: presto-main/src/main/java/com/facebook/presto/cost/CoefficientBasedStatsCalculator.java [hunk 14]
```diff
@@ -195,12 +180,12 @@ private PlanNodeStatsEstimate visitTableScanWithPredicate(TableScanNode node, Ex
...
+        private Constraint<ColumnHandle> getConstraint(TableScanNode node)
         {
             DomainTranslator.ExtractionResult decomposedPredicate = DomainTranslator.fromPredicate(
                     metadata,
                     session,
-                    predicate,
+                    BooleanLiteral.TRUE_LITERAL,
                     types);
 
             TupleDomain<ColumnHandle> simplifiedConstraint = decomposedPredicate.getTupleDomain()
```
[the remark this answers is not in the document]
Comment (findepi) [reply presto-main/src/main/java/com/facebook/presto/cost/CoefficientBasedStatsCalculator.java]: I don't understand. Anyway, are you sure such a change would belong to `Remove Filter above TableScan logic from CoefficientBasedStatsCalculator` commit?

File: presto-main/src/main/java/com/facebook/presto/cost/SymbolStatsEstimate.java (hunk 7 above)
Comment (findepi) [reply presto-main/src/main/java/com/facebook/presto/cost/SymbolStatsEstimate.java]: this didn't overflow line length limit, but i changed this nonetheless

File: presto-main/src/main/java/com/facebook/presto/cost/SymbolStatsEstimate.java (hunk 8 above)
Comment (findepi) [reply presto-main/src/main/java/com/facebook/presto/cost/SymbolStatsEstimate.java]: Good idea. I am concerned, however, that there might be some unexpected side effect to such change in formulae and we are unable to re-verify them yet. I've added a TODO comment instead.

File: presto-main/src/main/java/com/facebook/presto/cost/SymbolStatsEstimate.java (hunk 10 above)
Comment (findepi) [reply presto-main/src/main/java/com/facebook/presto/cost/SymbolStatsEstimate.java]: same as in `nullsFraction`

File: presto-tpch/src/main/java/com/facebook/presto/tpch/TpchMetadata.java (hunk 4 above)
Comment (findepi) [reply presto-tpch/src/main/java/com/facebook/presto/tpch/TpchMetadata.java]: i removed the commit from this PR.

Comment (findepi) [reply presto-tpch/src/main/java/com/facebook/presto/tpch/TpchMetadata.java]: oh, i didn't, it was a required commit
